add option "spawn in party" to adopt a hero

NathanGASC opened the pull request:
Want to add this option to be able to spawn heroes directly in the hero party.

For now only tested with & without config:
- create new
- spawn in party
- In player Clan
- In player Companion

Blocking issue when create new isn't used (can't find compatible heroes)

NathanGASC committed:
add an open api

NathanGASC committed:
Revert "add an open api"

This reverts commit c9311b659889f270eae8d38ee0aefdcf8b83f41e.

NathanGASC committed:
add spawn in main hero party option

NathanGASC committed:
add 2 for Adopt a hero & 1 global option

billw2012 commented:
Can you describe more what exactly this does please?

NathanGASC commented:
I add 3 options for adopt a hero.
- Join player clan : define if the viewer pnj should join the clan.
- Join player companion : define if the viewer pnj should join the companion list.
- Spawn in the party : define if the viewer pnj should spawn in party (if not, it will spawn in a random city)

I also add an option in Adopt a Hero - general config:
- Break companion limit: will put the companion limit to : actual companion count + 1. (This rule is when a streamer wants to play with a lot of viewer companion)

NathanGASC committed:
Merge branch 'billw2012:main' into main

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/Actions/AdoptAHero.cs:
```diff
@@ -240,6 +247,33 @@ private static (bool success, string message) ExecuteInternal(string args, strin
                 return (false, "You can't adopt a hero: no available hero matching the requirements was found!");
             }
             
+            if (settings.SpawnInParty)
```
Do you know what the implications are of adding a hero to party that isn't also a companion? It seems a bit weird, but I remember some quests do something like this so maybe its fine? But can the player *remove* them again? What happens if they do? This seems like something that needs some comprehensive testing!

NathanGASC replied:
Alone, this command don't have a lot of interest. If this command is the only selected one, you will not be able to remove them from you party, you will have to search for them in cities to get them back and you will have to buy them again.

But, it's interesting with other commands. However, some may find it too strong to spawn heroes in party and will want to disable it.

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/Actions/AdoptAHero.cs:
```diff
@@ -260,6 +294,19 @@ private static (bool success, string message) ExecuteInternal(string args, strin
+                }
+            }
+
+            if (settings.JoinPlayerCompanion)
```
Its possible to add as companion but NOT add to party I guess? Seems a bit weird, maybe `JoinPlayerCompanion` should imply join party? Otherwise they are at some random town still. But maybe that is good? I don't know...

NathanGASC replied:
I think it's a right idea to avoid option that will do some weird comportment happen with some configuration. In this way, you are right. Why don't setup a suboption for JoinPlayerCompanion which will be "in party". This way, i don't think there is weird comportment.

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/Actions/AdoptAHero.cs:
```diff
@@ -260,6 +294,19 @@ private static (bool success, string message) ExecuteInternal(string args, strin
 
+            if (settings.JoinPlayerClan)
+            {
+                if (Clan.PlayerClan.CompanionLimit < Clan.PlayerClan.Companions.Count)
```
Why does CompanionLimit matter when joining clan?

NathanGASC replied:
Fail from me. I should do this test when the newHero want to join companion, not the clan. We don't want to break the companion limit, except if the general option "break companion limit" is selected.

NathanGASC replied:
Removed join clan which have no effect on the game. You also have to be a companion or familly member (apparently) to be in a clan.

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/BLTAdoptAHero.cs:
```diff
@@ -66,6 +69,16 @@ public static void MissionNameMarkerVMConstructorPostfix(MissionNameMarkerVM __i
+        [UsedImplicitly, HarmonyPatch(typeof(DefaultClanTierModel), nameof(DefaultClanTierModel.GetCompanionLimit))]
+        public static void Postfix(ref int __result)
+        {
+            if (GlobalCommonConfig.Get() != null && GlobalCommonConfig.Get().BreakCompanionLimit)
```
Use `BLTAdoptAHeroModule.CommonConfig` instead, `.Get()` parses it from json each time.

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/BLTAdoptAHero.cs:
```diff
@@ -153,6 +166,11 @@ internal class GlobalCommonConfig
         public bool UseRawXP { get; set; } = true;
 
+        [Category("General"),
+        Description("Will disable companion limit. You will be able to"),
```
Description is incomplete

NathanGASC replied:
D:

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/BLTAdoptAHero.csproj:
```diff
@@ -36,19 +36,19 @@
     <ItemGroup>
         <Reference Include="0Harmony, Culture=neutral, PublicKeyToken=null">
-            <HintPath>..\..\..\..\..\..\..\Program Files (x86)\Steam\steamapps\common\Mount &amp; Blade II Bannerlord\Modules\Bannerlord.Harmony\bin\Win64_Shipping_Client\0Harmony.dll</HintPath>
+            <HintPath>C:\Program Files (x86)\Steam\steamapps\common\Mount &amp; Blade II Bannerlord\Modules\Bannerlord.Harmony\bin\Win64_Shipping_Client\0Harmony.dll</HintPath>
```
Update from main to get my project file changes

NathanGASC replied:
Yes, i will do that. Already updated it but not pushed yet. Nice change btw. It's make things easier ^^.

NathanGASC committed:
Merge branch 'main' into adopt_more_options

NathanGASC committed:
applie code review fix & remove join player clan option (which was useless)

NathanGASC commented:
I've updated the PR and apply your advices in the code review.
- removed join clan option (because you have to be a companion or familly member to be in the clan apparently)
- spawn in party will work only when join companion will be true.

NathanGASC committed:
Merge branch 'billw2012:main' into main

NathanGASC committed:
Merge branch 'main' into adopt_more_options

billw2012 committed:
Merge branch 'main' into adopt_more_options

billw2012 commented on BannerlordTwitch/BLTAdoptAHero/Actions/AdoptAHero.cs:
```diff
@@ -174,6 +178,33 @@ private static (bool success, string message) ExecuteInternal(string args, strin
+            }
+            else
+            {
+                var targetSettlement = Settlement.All.Where(s => s.IsTown).SelectRandom();
```
I missed it before, but this is wrong, the character should ONLY be put into a settlement if they were newly created, which won't be the case unless Create New is checked, so you can make this `else if(settings.CreateNew)`.
As it is currently it will move existing wanderers, and existing nobles into random settlements.

billw2012 reviewed (changes requested):
One issue remaining to be fixed, then I will merge it.

billw2012 committed:
Merge branch 'main' into adopt_more_options
